netplan/cli/utils.py

```diff
@@ -62,6 +65,40 @@ def systemctl_network_manager(action):  # pragma: nocover (covered in autopkgtes
     subprocess.check_call(['systemctl', action, '--no-block', service_name])
 
 
+def gather_replug_yaml(root_dir):
+    # create the file list
+    # per generate.c, /run/netplan shadows /etc/netplan/, which shadows /lib/netplan
+
+    names_to_paths = {}
+    for yaml_dir in ['lib', 'etc', 'run']:
+        for yaml_file in glob.glob(os.path.join(root_dir, yaml_dir, 'netplan', '*.yaml')):
+            names_to_paths[os.path.basename(yaml_file)] = yaml_file
+
+    files = [names_to_paths[name] for name in sorted(names_to_paths.keys())]
+
+    # now build the result from each file
+    result = {'disable_all_replug': False, 'blacklist': []}
+    for yaml_file in files:
+        try:
+            with open(yaml_file) as f:
+                yaml_data = yaml.load(f)
+        except (IOError, yaml.YAMLError):
+            logging.error('Error while loading %s, aborting.' % yaml_file)
+            sys.exit(1)
+
+        if 'replug' not in yaml_data:
+            continue
+
+        yaml_data = yaml_data['replug']
+        if 'disable_all_replug' in yaml_data:
+            result['disable_all_replug'] = yaml_data['disable_all_replug']
```

Comment:
I'm not convinced yet that disable_all_replug is a good idea even if it might be convenient to users. That was also the point of having the whole replug story elsewhere.

If we're going to change something, I feel like we should instead have a single keyword that allows skipping replug on a per-interface basis.

The essential idea I have is that you'd want to not recommend users disable replug -- in general, it's a good idea to have the replugging happen, and not doing it can break renaming interfaces; let's not make it hard for people to disable it, but let's also make sure it's absolutely a conscious decision.

Reply:
My concern about making it per-interface in this way how that would interact with the matching rules.

I think it's a bit chicken-and-egg. Looking at `apply.py`:
 - the backends get to try to bring up interfaces (apply.py, line 60)
 - then all interfaces which are down are replugged
 - if anything was replugged, udev is settled and the backends try again.

How does step 2 match an interface? Clearly it wasn't matched in step 1, otherwise the interface would be up - so it isn't matching on MAC address. Matching by current name won't work as the point of replugging is that you cannot rely on the original name, and if you're matching by the new name then that won't take effect until after replug, by which time the device will have failed. So I think this would only work if you're matching by driver - I can't think of anything else that would serve this purpose.

This means that if we have a user who has a device that is going away on replug, we first have to tell them to match by driver, and then add a stanza to stop that being replugged. (Also, if they have multiple devices like this - e.g. multiple Xen vifs or multiple qeths, which is very possible - they have to add the driver match and the replug prohibition to each one.)

This will also make it impossible to support the existing hardcoded blacklist in the same way as we support user-configured blacklists, which I think is a cute feature.